### PR TITLE
Checkpointing and retrieval

### PR DESCRIPTION
**Pull request by NicolasRR**
Implemented checkpoint and retrieval for model, scheduler, and optimizer state_dict as well as random generator states for reproducibility.

### CODE REVIEW

File: src/optim/base.py [hunk 1]
```diff
@@ -50,6 +52,12 @@ def train_base(model, opt, data, data_seed, scheduler, iterations, acc_steps, ba
     t0 = time.time()
     train_epochs = 0
     while itr < iterations:
+        if not rng_state_dict is None:
```
Comment (mkrima) [inline src/optim/base.py:55]: this should only happen in the first execution of the while and not in the next ones. can you move it outside the while?

File: src/main.py [hunk 2]
```diff
@@ -112,6 +112,26 @@ def main(args):
...
+
+        optimizer_state_dict = checkpoint['optimizer']
+        scheduler_state_dict = checkpoint['scheduler']
+        rng_state_dict = {
+            module: checkpoint[module] for module in ["cpu_rng_state", "gpu_rng_state", "numpy_rng_state", "py_rng_state"]
+        }
+
+        model.load_state_dict(model_state_dict) 
+        opt.load_state_dict(optimizer_state_dict)
+        scheduler.load_state_dict(scheduler_state_dict)
```
Comment (mkrima) [inline src/main.py:133]: maybe check here if scheduler is not None. (also related to #8)

File: src/main.py [hunk 3]
```diff
@@ -112,6 +112,28 @@ def main(args):
...
+
+        optimizer_state_dict = checkpoint['optimizer']
+        rng_state_dict = {
+            module: checkpoint[module] for module in ["cpu_rng_state", "gpu_rng_state", "numpy_rng_state", "py_rng_state"]
+        }
+
+        model.load_state_dict(model_state_dict) 
+        opt.load_state_dict(optimizer_state_dict)
+        itr=checkpoint['itr']
+        if not scheduler is None:
```
Comment (mkrima) [inline src/main.py:134]: scheduler is not None

File: src/main.py [hunk 4]
```diff
@@ -112,6 +112,28 @@ def main(args):
     elif os.path.isfile(os.path.join(ckpt_path, "summary.json")): # the experiment was already completed
         print(f"Already found experiment '{ckpt_path}'.\nSkipping.")
         sys.exit(0)
+    itr = 0
+    rng_state_dict = None
+    distributed_backend.sync()
```
Comment (mkrima) [inline src/main.py:117]: can you put this inside the if not os.path.exists(ckpt_path)? it makes it more clear why it's needed.

Comment (NicolasRR) [reply src/main.py:117]: I only moved   ` distributed_backend.sync()` inside as the other variables need to be defined if the path already exists but there are no checkpoints which might happen in case of crashing early on

Comment (mkrima) [reply src/main.py]: yes that's exactly what I meant. Thank you.

**Comment by mkrima** [comment]
LGTM

**Comment by haeggee** [comment]
quick question: without having gone through the code in detail, does this also make sure the data sampling is deterministic (dataloader state is restored)? what i mean is, does the model see the same data when loading from a checkpoint vs when doing a full training run

**Comment by NicolasRR** [comment]
Good point, I have added those modifications in another PR
